File: java/src/org/revenj/postgres/converters/StringConverter.java
package org.revenj.postgres.converters;

import org.revenj.postgres.PostgresReader;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public abstract class StringConverter {
	public static void skip(PostgresReader reader, int context) throws IOException {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return;
		}
		if (cur != '"' && cur != '\\') {
			reader.initBuffer();
			reader.fillUntil(',', ')');
		} else {
			cur = reader.read(context);
			while (cur != -1) {
				if (cur == '\\' || cur == '"') {
					cur = reader.read(context);
					if (cur == ',' || cur == ')') {
						return;
					}
					cur = reader.read(context);
				} else cur = reader.read();
			}
			throw new IOException("Unable to find end of string");
		}
	}

	public static String parse(PostgresReader reader, int context) throws IOException {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return null;
		}
		if (cur != '"' && cur != '\\') {
			reader.initBuffer((char) cur);
			reader.fillUntil(',', ')');
			reader.read();
			return reader.bufferToString();
		}
		return parseEscapedString(reader, context, cur, ')');
	}

	private static String parseEscapedString(PostgresReader reader, int context, int cur, char matchEnd) throws IOException {
		cur = reader.read(context);
		reader.initBuffer();
		do {
			if (cur == '\\' || cur == '"') {
				cur = reader.read(context);
				if (cur == ',' || cur == matchEnd) {
					return reader.bufferToString();
				}
				for (int i = 0; i < context - 1; i++) {
					cur = reader.read();
				}
			}
			reader.addToBuffer((char) cur);
			reader.fillUntil('\\', '"');
			cur = reader.read();
		} while (cur != -1);
		throw new IOException("Unable to find end of string");
	}

	public static List<String> parseCollection(PostgresReader reader, int context, boolean allowNull) throws IOException {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return null;
		}
		boolean espaced = cur != '{';
		if (espaced) {
			reader.read(context);
		}
		int innerContext = context << 1;
		cur = reader.peek();
		if (cur == '}') {
			if (espaced) {
				reader.read(context + 2);
			} else {
				reader.read(2);
			}
			return new ArrayList<>(0);
		}
		List<String> list = new ArrayList<>();
		String emptyCol = allowNull ? null : "";
		do {
			cur = reader.read();
			if (cur == '"' || cur == '\\') {
				list.add(parseEscapedString(reader, innerContext, cur, '}'));
				cur = reader.last();
			} else {
				reader.initBuffer((char) cur);
				reader.fillUntil(',', '}');
				cur = reader.read();
				if (reader.bufferMatches("NULL")) {
					list.add(emptyCol);
				} else {
					list.add(reader.bufferToString());
				}
			}
		} while (cur == ',');
		if (espaced) {
			reader.read(context + 1);
		} else {
			reader.read();
		}
		return list;
	}

	public static int serializeCompositeURI(String value, char[] buf, int pos) {
		for (int i = 0; i < value.length(); i++) {
			char c = value.charAt(i);
			if (c == '\\' || c == '/') {
				buf[pos++] = '\\';
			}
			buf[pos++] = c;
		}
		return pos;
	}
}

File: csharp/Core/Revenj.Core/DatabasePersistence/Postgres/Converters/StringConverter.cs
using System.Collections.Generic;
using System.IO;
using Revenj.Common;
using Revenj.Utility;

namespace Revenj.DatabasePersistence.Postgres.Converters
{
	public static class StringConverter
	{
		public static void Skip(BufferedTextReader reader, int context)
		{
			var cur = reader.Read();
			if (cur == ',' || cur == ')')
				return;
			if (cur != '"' && cur != '\\')
			{
				reader.InitBuffer();
				reader.FillUntil(',', ')');
				reader.Read();
			}
			else
			{
				cur = reader.Read(context);
				while (cur != -1)
				{
					if (cur == '\\' || cur == '"')
					{
						cur = reader.Read(context);
						if (cur == ',' || cur == ')')
							return;
						cur = reader.Read(context);
					}
					else cur = reader.Read();
				}
				throw new FrameworkException("Unable to find end of string");
			}
		}

		public static string Parse(BufferedTextReader reader, int context)
		{
			var cur = reader.Read();
			if (cur == ',' || cur == ')')
				return null;
			if (cur != '"' && cur != '\\')
			{
				reader.InitBuffer((char)cur);
				reader.FillUntil(',', ')');
				reader.Read();
				return reader.BufferToString();
			}
			return ParseEscapedString(reader, context, ref cur, ')');
		}

		private static string ParseEscapedString(BufferedTextReader reader, int context, ref int cur, char matchEnd)
		{
			cur = reader.Read(context);
			reader.InitBuffer();
			do
			{
				if (cur == '\\' || cur == '"')
				{
					cur = reader.Read(context);
					if (cur == ',' || cur == matchEnd)
						return reader.BufferToString();
					for (int i = 0; i < context - 1; i++)
						cur = reader.Read();
				}
				reader.AddToBuffer((char)cur);
				reader.FillUntil('\\', '"');
				cur = reader.Read();
			} while (cur != -1);
			throw new FrameworkException("Unable to find end of string");
		}

		public static Stream ParseStream(BufferedTextReader reader, int context)
		{
			var cur = reader.Read();
			if (cur == ',' || cur == ')')
				return null;
			var cms = ChunkedMemoryStream.Create();
			var sw = cms.GetWriter();
			if (cur != '"' && cur != '\\')
			{
				sw.Write((char)cur);
				reader.FillUntil(sw, ',', ')');
			}
			else
			{
				cur = reader.Read(context);
				while (cur != -1)
				{
					if (cur == '\\' || cur == '"')
					{
						cur = reader.Read(context);
						if (cur == ',' || cur == ')')
						{
							sw.Flush();
							cms.Position = 0;
							return cms;
						}
						for (int i = 0; i < context - 1; i++)
							cur = reader.Read();
					}
					sw.Write((char)cur);
					cur = reader.Read();
				}
				reader.Read(context);
			}
			sw.Flush();
			cms.Position = 0;
			return cms;
		}

		public static List<string> ParseCollection(BufferedTextReader reader, int context, bool allowNull)
		{
			var cur = reader.Read();
			if (cur == ',' || cur == ')')
				return null;
			var espaced = cur != '{';
			if (espaced)
				reader.Read(context);
			cur = reader.Peek();
			if (cur == '}')
			{
				if (espaced)
					reader.Read(context + 2);
				else
					reader.Read(2);
				return new List<string>(0);
			}
			var innerContext = context << 1;
			var list = new List<string>();
			var emptyCol = allowNull ? null : string.Empty;
			do
			{
				cur = reader.Read();
				if (cur == '"' || cur == '\\')
					list.Add(ParseEscapedString(reader, innerContext, ref cur, '}'));
				else
				{
					reader.InitBuffer((char)cur);
					reader.FillUntil(',', '}');
					cur = reader.Read();
					if (reader.BufferMatches("NULL"))
						list.Add(emptyCol);
					else
						list.Add(reader.BufferToString());
				}
			} while (cur == ',');
			if (espaced)
				reader.Read(context + 1);
			else
				reader.Read();
			return list;
		}

		private static readonly byte[] NULL = new byte[] { (byte)'N', (byte)'U', (byte)'L', (byte)'L' };

		public static List<Stream> ParseStreamCollection(BufferedTextReader reader, int context, bool allowNull)
		{
			var cur = reader.Read();
			if (cur == ',' || cur == ')')
				return null;
			var espaced = cur != '{';
			if (espaced)
				cur = reader.Read(context);
			var innerContext = context << 1;
			var list = new List<Stream>();
			cur = reader.Peek();
			if (cur == '}')
				reader.Read();
			while (cur != -1 && cur != '}')
			{
				cur = reader.Read();
				var cms = ChunkedMemoryStream.Create();
				var sw = cms.GetWriter();
				if (cur == '"' || cur == '\\')
				{
					cur = reader.Read(innerContext);
					while (cur != -1)
					{
						if (cur == '\\' || cur == '"')
						{
							cur = reader.Read(innerContext);
							if (cur == ',' || cur == '}')
								break;
							for (int i = 0; i < innerContext - 1; i++)
								cur = reader.Read();
						}
						sw.Write((char)cur);
						cur = reader.Read();
					}
					sw.Flush();
					cms.Position = 0;
					list.Add(cms);
				}
				else
				{
					do
					{
						sw.Write((char)cur);
						cur = reader.Read();
					} while (cur != -1 && cur != ',' && cur != '}');
					sw.Flush();
					cms.Position = 0;
					if (cms.Matches(NULL))
					{
						list.Add(null);
						cms.Dispose();
					}
					else
						list.Add(cms);
				}
			}
			if (espaced)
				reader.Read(context + 1);
			else
				reader.Read();
			return list;
		}

		public static int SerializeCompositeURI(string value, char[] buf, int pos)
		{
			foreach (var c in value)
			{
				if (c == '\\' || c == '/')
					buf[pos++] = '\\';
				buf[pos++] = c;
			}
			return pos;
		}
	}
}
